RFC SPI: Reference Implementation.

ithinuel opened the pull request:
### Description

This PR contains reference implementations of the SPI HAL RFC changes for the FRDM-K66F and NUCLEO_F429ZI boards.

#### Known issues :
- The implementation for the K66F does not support least significant bit first in slave mode. This is due to a hardware limitation that needs to be workedaround in software.

### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [x] Breaking change

ithinuel committed:
update HAL API

cmonr commented:
_sigh_

I think the feature branch needs to be updated.

cmonr commented:
Ooor maybe this PR needs work? I dunno.

The specific failure in cloud-client-test:
> [Build K66F ARM] [Error] PN512SPITransportDriver.cpp@40,0:  #308: more than one instance of overloaded function "mbed::SPI::frequency" matches the argument list:
[Build K66F ARM] [ERROR] "./mbed-os/features/nfc/source/controllers/PN512SPITransportDriver.cpp", line 40: Error:  #308: more than one instance of overloaded function "mbed::SPI::frequency" matches the argument list:
[Build K66F ARM]             function "mbed::SPI::frequency(int)"
[Build K66F ARM]             function "mbed::SPI::frequency(std::uint32_t)"
[Build K66F ARM]             argument types are: (unsigned long)
[Build K66F ARM]             object type is: mbed::SPI
[Build K66F ARM] ./mbed-os/features/nfc/source/controllers/PN512SPITransportDriver.cpp: 0 warnings, 1 error

ithinuel commented:
> /builds/ws/mbed-os-ci_unittests/unitTest-385/mbed-os/features/netsocket/SocketAddress.cpp:22:23: fatal error: ip4string.h: No such file or directory

we might need to rerebase the feature branch indeed.

> The command "python tools/make.py -t GCC_ARM -m K82F --source=. --build=BUILD/K82F/GCC_ARM -j0" exited with 1.

and 

> The command "python tools/make.py -t GCC_ARM -m K64F --source=. --build=BUILD/K64F/GCC_ARM -j0" exited with 1.

These are expected failures because the port of the driver is only valid for the FRDM_K66F and the NUCLEO_F429ZI.

c1728p9 commented on drivers/SPI.cpp:
```diff
@@ -16,97 +16,146 @@
 #include "platform/mbed_critical.h"
 
-#if DEVICE_SPI_ASYNCH
+#if 0
```
Why is this code disabled?

ithinuel replied:
This is because the async API in the driver is not very clear on :
- Locks :
  - Shall lock match the lock of the peripheral (thread safety of the peripheral)
  - Shall lock match the lock of the driver instance to race (thread safety of the instance)
- chip select behaviours
  - How should that behave when chip select is handled in hardware ?
  - What is the relation between the chip select and the locks ?

c1728p9 commented on drivers/SPI.cpp:
```diff
@@ -16,97 +16,146 @@
+    struct SPI::spi_peripheral_s *result = NULL;
+    core_util_critical_section_enter();
+    for (uint32_t idx = 0; idx < SPI_COUNT; idx++) {
+        printf("SPI::lookup(%08x) found at %lu", name, idx);
```
printf should be removed

ithinuel replied:
fixed

c1728p9 commented on drivers/SPI.cpp:
```diff
@@ -16,97 +16,146 @@
     _acquire();
-    int ret = spi_master_write(&_spi, value);
+    uint32_t ret = 0;
+    spi_transfer(&_self->spi, &value, _bits/8, &ret, _bits/8, NULL);
```
You'll probably want to round up here.

ithinuel replied:
Indeed ! Fixed :)

c1728p9 commented on drivers/SPISlave.cpp:
```diff
@@ -45,17 +45,17 @@ void SPISlave::frequency(int hz)
 int SPISlave::receive(void)
 {
-    return (spi_slave_receive(&_spi));
+    return 0;//(spi_slave_receive(&_spi));
```
Why is function commented out? Was the intent to remove it?

ithinuel replied:
It looks like I forgot the commit SPISlave driver changes. This should be ok now.

c1728p9 commented on targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_K66F/TARGET_FRDM/PeripheralPins.c:
```diff
@@ -185,7 +185,7 @@ const PinMap PinMap_SPI_SCLK[] = {
 };
 
-const PinMap PinMap_SPI_MOSI[] = {
+const PinMap PinMap_SPI_SOUT[] = {
```
Why was this changed from MOSI to SOUT? Same for MISO to SIN below.

ithinuel replied:
This is because on Freescale's targets the pins aren't MISO & MOSI. If you switch from master to slave, the direction of the pin will not be changed to match it's new role.
On Freescale's targets, the SIN is always the input what ever the mode (master or slave) of the peripheral.

deepikabhavnani commented on drivers/SPI.cpp:
```diff
@@ -16,97 +16,144 @@
 }
 
-void SPI::frequency(int hz)
+uint32_t SPI::frequency(uint32_t hz)
```
Variable/Parameter types should be consistent in driver. Existing driver uses `int`, `char`.. etc, changed here for few API's to `uint32_t` `uint8_t` and not all.  I would expect API's to be consistent in a driver. 

Also if we are changing the return type, should we deprecate the existing one?  In case of `format` for addition of new parameter we are adding new API, but old is not deprecated and for `frequency` we are changing the return type and parameter type of API without any deprecation.

ithinuel replied:
The changes to this driver are provisional and meant to be minimal.
Any change to bring consistency may break existing API.

We cannot deprecate the void returning version of `SPI::frequency` because return types are not "overloadable". Also, going from `void` to `uint32_t` do not present risk on ignoring this return value.

kjbracey commented on drivers/SPI.h:
```diff
@@ -74,9 +77,14 @@ namespace mbed {
+    struct spi_peripheral_s {
+        SPIName name;
+        spi_t spi;
+        PlatformMutex *mutex;
```
Would it not be easier to make this a `SingletonPtr<PlatformMutex>`?

kjbracey replied:
That's probably pushing too much stuff into the static allocation though. After more thought, this seems like a bad hybrid. There is a benefit to being fully-static, but if you're going to start doing some dynamic allocation, it may as well be fully dynamic to minimise overhead. I think rather than this static array, it would be better just to have the head of a linked list of peripherals, each of which with an embedded `PlatformMutex` and `spi_t`, so one heap block per peripheral.

ithinuel replied:
Ideally I would want the PlatformMutex to be statically allocated but AFAIU this is causing issues with the constructor not being invoked when instantiated from a static array..

kjbracey commented on drivers/SPI.h:
```diff
@@ -268,42 +290,34 @@ class SPI : private NonCopyable<SPI> {
+    // Drawback: it costs ram size even if the device is not used.
+    static spi_peripheral_s _peripherals[SPI_COUNT];
+
+    spi_peripheral_s *_self;
```
`_self` as a name is far too vague - it should surely be `_peripheral`.

ithinuel replied:
This is a little bikeshedding but I picked _self because :
- `_peripheral` do not reflect enough that this member is the peripheral associated to `this` SPI instance.
- `_this` would have been ambiguous with `this`.

kjbracey replied:
> `_peripheral` do not reflect enough that this member is the peripheral associated to this SPI instance.

I'm not sure how it could be any clearer. It tells us it's a/the peripheral, and obviously it's associated with this `SPI` instance, because our `SPI` object pointing to it. It's `this->_peripheral`.

`SPI` is the object representing an SPI slave we're talking to, and then we have a pointer to the peripheral controlling it.

`this->_self` reads to me just the same as saying `this->_this`.

kjbracey commented on drivers/SPI.h:
```diff
@@ -74,9 +77,14 @@ namespace mbed {
  */
 class SPI : private NonCopyable<SPI> {
-
+protected:
```
Why protected? Is this intentionally exposed as API?

ithinuel replied:
Because the member variables using this type are protected too.

kjbracey replied:
`_self`? Does that need to be protected rather than private? You want to permit derived classes to mess with it? I'd be inclined to keep it private until there's a definite need to reveal it.

(Watching all the "protected exclusion" markers being applied for Doxygen recently has brought home to me how much excess protected there is. `protected` is often unintentionally exposed API. Either it shouldn't be protected, or it should be documented (as `@internal` if necessary.)

kjbracey commented on drivers/SPI.cpp:
```diff
@@ -16,97 +16,144 @@
+    // unless not assigned to a peripheral, this mutex will never be deleted.
+    PlatformMutex *mutex = new PlatformMutex();
+
+    core_util_critical_section_enter();
```
Critical section seems heavy, and I'm wondering if that's tying your hands from using `SingletonPtr` for the `PlatformMutex`.

You don't need your own mutex, you could just borrow `singleton_lock` from `SingletonPtr`, which is used by that and all other static initialisation.

kjbracey commented on drivers/SPI.cpp:
```diff
@@ -16,97 +16,144 @@
+    core_util_critical_section_enter();
+    for (uint32_t idx = 0; idx < SPI_COUNT; idx++) {
+        if ((_peripherals[idx].name == name) ||
+            (_peripherals[idx].name == 0)) {
```
`(_peripherals[idx].name == 0 && or_last)` would void the need for lines 84-86.

ithinuel replied:
Considering that `SPI_COUNT` will always be small, it's indeed not a big cost of running the loop even if we already reached a 0'ed name.
Thinking about that, 0 might not be the best choice as the reserved value for invalid name is NC and is usually -1.

kjbracey replied:
Not really thinking about performance - just about code clarity - this is a complicated way of matching "target, or optionally empty" that relies on the special case of all empties being at the end. Expressing the condition in the loop as "name == target or optionally empty" does what you want immediately.

Main cost I'm worried about is the RAM size of the array, which is why I later suggested a linked list, if you're resigned to some dynamic allocation anyway.

kjbracey commented on drivers/SPI.cpp:
```diff
@@ -16,97 +16,144 @@
+            result = &_peripherals[idx];
+            break;
+        }
+        // XXX: we may want to ensure that it was previously initialized with the same mosi/miso/sclk/ss pins
```
I would certainly be inclined to have an assert for this. If you do see the same peripheral with different pins, you could in principle multiplex the peripheral across the two pin sets by recalling `spi_init` on every acquire, but I suspect this is not a needed use case. Currently no other peripheral I'm aware of can cope with that scenario, and I think they generally fail quietly because of no proper detection (HAL quietly accepts successive xxx_init calls).

Wouldn't want to store extra info to monitor this, but I guess we can cheat and approximate by looking at the existing owner, if any, and checking their pins.

ithinuel replied:
I have added elements to allow that. It costs 4 PinNames per peripheral.

kjbracey commented on drivers/SPI.cpp:
```diff
@@ -16,97 +16,144 @@
+    if (_self->owner == this) {
+        spi_format(&_self->spi, _bits, _mode, _msb_first?SPI_BIT_ORDERING_MSB_FIRST:SPI_BIT_ORDERING_LSB_FIRST);
     } else {
         _acquire();
```
I don't believe you actually need to acquire here, although the frequency case now does for its "actual" value.

kjbracey commented on drivers/SPI.cpp:
```diff
@@ -16,97 +16,144 @@
 }
 
-void SPI::frequency(int hz)
+uint32_t SPI::frequency(uint32_t hz)
```
I don't see a problem extending by adding a return to a void function, but not sure what the point of changing the input type is. That is a binary compatibility break. (Adding a scalar return isn't, afaik).

kjbracey commented on drivers/SPI.cpp:
```diff
@@ -115,29 +162,29 @@ int SPI::write(const char *tx_buffer, int tx_length, char *rx_buffer, int rx_len
 {
     lock();
-    _write_fill = data;
+    _write_fill = (uint32_t)data;
```
What is this cast doing? If anything you should be widening the input type here.

kjbracey commented on drivers/SPI.h:
```diff
@@ -103,12 +111,13 @@ class SPI : private NonCopyable<SPI> {
      * @endcode
      */
     void format(int bits, int mode = 0);
+    void format(uint8_t bits, spi_mode_t mode = SPI_MODE_IDLE_LOW_SAMPLE_FIRST_EDGE, bool msb_first = true);
```
I guess we probably do want to retain the original method for binary compatibility, but we could change the `0` to the define.

As for `msb_first` - you've got an odd mixture of `bool` and defines/enum.  Can't it be consistently be one or the other?

ithinuel replied:
Good point, I'll fallback to the `spi_bit_ordering_t` enum as it is the type used inside the implementation.

kjbracey commented on drivers/SPI.cpp:
```diff
@@ -16,97 +16,144 @@
+    }
+    if (_self->name == 0) {
+        _self->name = name;
+        spi_init(&_self->spi, false, mosi, miso, sclk, NC);
```
`ssel` is now being totally ignored. I still think that's a retrograde step - it is notionally within the bounds of existing platform-specific behaviour, but it would surely be better to have `lock` manipulate `ssel` as a GPIO.

AFAIK, we're already in the happy state where all portable code already passes NC as ssel and does it manually because of the variable behaviour of platforms, so we now have the room to start making it work.

kjbracey replied:
On further reflection, probably best to not put it in actual `lock`, as we do have a common pattern where `lock` is just a "peripheral lock", and it's virtual so that people can inherit and override, in particular for bypassing the mutex for use from interrupt, if they know they're the only user of the peripheral.

There are some `UnlockedSPI` out there that inherit from `SPI` for use from IRQ, overriding `lock`+`unlock`. Shouldn't break that pattern.

Instead, having `select` and `deselect` methods that are `lock; acquire; cs=1 (reference counted)` would be the way to go - that would still work when someone overrides lock to eliminate the peripheral mutex.

ithinuel replied:
My bad, this NC should be ssel.

Any work regarding locks/selections is out of scope from this PR though.

kjbracey replied:
It has to be `NC` - can't be anything else or the entire "peripheral" concept won't work. That inherently prevents any use of the HAL ssel, because you're sharing `spi_t` for the same bus.

kjbracey replied:
Well, I guess you /could/ pass ssel to the HAL, but only if you did it on every acquire. But that means a full `init` every time you change owner of a peripheral, I guess?

kjbracey replied:
> Any work regarding locks/selections is out of scope from this PR though.

From my fairly prolonged personal experience of using the SPI API, the two biggest problems are the system-wide transfer mutex, and the total uselessness of the SSEL parameter.   (Lack of trust of the async API implementations would be a runner-up)

Great that you're fixing the first one - which is also a purely C++ layer and non-HAL thing - but I don't see how you can do any sort of SPI rework that doesn't address the SSEL problem. If it's not going to be addressed when during an SPI API rework, when is it going to be addressed?

The peripheral change as you've written it here is effectively is going to force some SSEL change. You're going to have to do something. Any of these are possible:

* render SSEL totally non-functional for all platforms, as the current PR, everyone does SSEL manually via GPIO, as portable code already does anyway.
* rework the acquire mechanism to do `spi_init` to change SSEL
* give up the shared `spi_t`
* put the GPIO for SSEL inside `SPI` - doesn't break anyone with non-portable code relying on the SPI SSEL, and the portable code can be cleaned up to take out the manual handling.

I think the last is by far the best combination of easy+beneficial.

kjbracey commented on drivers/SPI.cpp:
```diff
@@ -16,97 +16,144 @@
     _acquire();
-    int ret = spi_master_write(&_spi, value);
+    uint32_t ret = 0;
+    spi_transfer(&_self->spi, &value, (_bits+7)/8, &ret, (_bits+7)/8, NULL);
```
This only works little-endian. If attempting portability you need a 3-way switch using `uint8_t`, 1;`uint16_t`, 2; or `uint32_t`, 4.

kjbracey replied:
Although, do we ever care? Not sure mbed OS supports big-endian at all, or have any plans to do so. It's just that the HAL api said something about platform endianness, suggesting coping with either.

ithinuel replied:
IMHO this method should be deprecated as it encourages reading symbols per symbol which is really bad for performances.
This fallback is provided for compatibility.

0xc0170 replied:
I assume this will be second step - another PR (deprecation note for this byte oriented methods) ?

kjbracey replied:
Wouldn't disagree with this - it's a convenient short-hand, but other APIs like `Socket` and `FileHandle` only give you the block form. If people really need byte-wise stuff, they can probably wrap a helper up themselves.

kjbracey commented on drivers/SPI.cpp:
```diff
@@ -115,29 +162,29 @@ int SPI::write(const char *tx_buffer, int tx_length, char *rx_buffer, int rx_len
     lock();
     _acquire();
-    int ret = spi_master_block_write(&_spi, tx_buffer, tx_length, rx_buffer, rx_length, _write_fill);
+    int ret = spi_transfer(&_self->spi, tx_buffer, tx_length, rx_buffer, rx_length, &_write_fill);
```
Again, this use of _write_fill assumes little endian. Possibly simplest to have a 3-way `uint8_t`/`uint16_t`/`uint32_t` union to store it.

kjbracey commented on drivers/SPISlave.cpp:
```diff
@@ -43,19 +47,34 @@ void SPISlave::frequency(int hz)
+        spi_transfer_async_abort(&_spi);
+        _is_pending = false;
+    }
+    _is_pending = spi_transfer_async(&_spi, &value, 1, (void *)&_buffer, 1, &_dummy, &SPISlave::irq_handler, this, (DMAUsage)0);
```
More illegal-if-big-endian stuff.

kjbracey commented on drivers/SPI.h:
```diff
@@ -268,42 +290,34 @@ class SPI : private NonCopyable<SPI> {
 protected:
-    spi_t _spi;
+    // holds spi_peripheral_s per peripheral on the device.
+    // Drawback: it costs ram size even if the device is not used.
```
This shouldn't be costing space if SPI isn't used at all, I hope. I guess we can live with using more than necessary for buses that aren't used.

kjbracey reviewed (changes requested):
The bus-separating logic here will be pretty vital - we're currently hitting problems with SD card traffic interfering with 802.15.4 radios on a different bus, and this will solve it.

(I had knocked up a quick hack to confirm this, and came back to check what the state was with this PR).

kjbracey reviewed (approved):
Okay, all my issues on this are to do with the state of C++ `SPI` class - I've got no issue with the HAL specification or implementation. As far as I can see all the tools are in place to get `SPI` working the way I think it should be. (And that would include not actually using the HAL SSEL, to have maximum flexibility for multiple devices on one bus). I believe it should also be possible to resurrect async with proper peripheral ownership management with current HAL API.

Therefore I'll approve this to allow HAL development to proceed, but more work on `SPI` and its APIs will be required before going to master. I will try to find time to put forward my own `SPI`-refining PR for this feature branch.

0xc0170 commented:
Thanks @kjbracey-arm ! Looking forward for SPI-refining

ithinuel committed:
update drivers

ithinuel committed:
reference implementation for K66F

Co-authored-by: Wilfried Chauveau <wilfried.chauveau@arm.com>
Co-authored-by: Przemyslaw Stekiel <przemyslaw.stekiel@mobica.com>

ithinuel committed:
reference implementation for NUCLEO_F429ZI

Co-authored-by: Wilfried Chauveau <wilfried.chauveau@arm.com>
Co-authored-by: Przemyslaw Stekiel <przemyslaw.stekiel@mobica.com>

0xc0170 commented:
@ithinuel The branch was recently rebased, can you update this one (there's one conflict as well now)

ithinuel committed:
disable SPI for all targets but K66F & NUCLEO_F429ZI

0xc0170 commented:
This shall be ready for CI but travis still fails . @ithinuel Please review the test failures (some look related)

ithinuel commented:
@0xc0170 
`travis-ci/events` and `travis-ci/littlefs` failures are due to the unsupported target K82F and K64F, the reference implementation being K66F.
`continuous-integration/travis-ci/pr` is a combination of the two previous failures.

cmonr commented:
@ithinuel One way of moving forward could be to disable those two tests in the feature branch, until they are ported.

ithinuel commented:
@cmonr For that purpose, is commenting out the corresponding lines in `.travis.yml` a valid solution ?

mprse committed:
Disable SPI for all targets other than K66F and NUCLEO_F429ZI.

Disable also components which uses SPI like SD, SPIF.

mprse committed:
Fix TARGET_MCUXpresso_MCUS/api/objects.h

Include "spi_api.h" only if DEVICE_SPI is defined (not DEVICE_SPI_ASYNCH).
Provide struct spi_s only if DEVICE_SPI is defined (uses types from "spi_api.h").

cmonr commented:
@ithinuel Yup!

Reference the currently commented out Py3.x tests that I still need to get validated and stable...

mprse committed:
Disable littlefs in .travis.yml

mprse commented:
`travis-ci/littlefs` still needs to be disabled. Fix can be found here: PR https://github.com/ithinuel/mbed-os/pull/7.

mprse commented:
Hooray! It looks like we can start CI.

0xc0170 reviewed (approved):
Note, this branch should be rebased - I can see astyle should fail for some changes, thus any new PR should already have latest travis to catch them.

0xc0170 commented:
CI started

0xc0170 commented:
The PR needs fixing (build failures)

mprse commented:
I checked build failure for `GCC/ARCH_PRO`:
`[Error] spi_api.c@56,31: unknown type name 'spi_t'`

It looks like spi target specific driver should be compiled only if `DEVICE_SPI` symbol is defined. I'll try to add the missiong preprocesor condition to all `spi_api.c` files.

Commit:
Add missing ifdef guards in spi_api.c files

Commit:
Add more DEVICE_SPI guards to other files

Commit:
Fix SPI drivers compatibility issues with Mbed 2

Commit:
And again more DEVICE_SPI guards

Commit:
Disable testing with examples using SPI for unsupported targets

Commit:
Fixed illegal inline member initialization in SPI.h

Commit:
An empty array in examples.json means no filter - urgh

Commit:
Fix if macro error in targets/TARGET_Silicon_Labs/TARGET_EFM32/TARGET_EFM32GG11/PeripheralPins.c

Commit:
Remove support for 802.15.4 PHY from KW24D as it is connected using SPI within the SoC

deepikabhavnani commented on targets/TARGET_Silicon_Labs/TARGET_EFM32/TARGET_EFM32PG/PeripheralPins.c:
```diff
@@ -198,6 +198,7 @@ MBED_WEAK const PinMap PinMap_PWM[] = {
 };
 
 /*************SPI**************/
+#if SPI_DEVICE
```
Should it be `DEVICE_SPI`?

donatieng replied:
Good catch, thanks @deepikabhavnani!

deepikabhavnani replied:
@donatieng - All the `#if` changes are relevant for master branch as well, it will be good to move common changes to master branch instead `feature` branch to have minimal conflicts in future, just a suggestion not blocker.

Commit:
Fix SPI_DEVICE <> DEVICE_SPI

mprse committed:
K64F, K66F: Remove "FLASHIAP" component.

When "SD" component is unavailable the following tests try to use Flash as storage and fail:
features-storage-tests-blockdevice-general_block_device
features-device_key-tests-device_key-functionality
features-storage-tests-kvstore-static_tests
features-storage-tests-kvstore-general_tests_phase_1
features-storage-tests-kvstore-securestore_whitebox
tests-psa-prot_internal_storage

Additional investigation is required to find the reason of the failures.

mprse commented:
Analysis of the failing tests:
https://mbed-os.mbedcloudtesting.com/blue/organizations/jenkins/mbed-os-ci_greentea-test/detail/mbed-os-ci_greentea-test/658/tests/

List of the failed tests below:

Test | Board | Compiler
-- | -- | --
features-storage-tests-blockdevice-general_block_device | K82F, K64F | ARM, GCC_ARM, IAR
features-device_key-tests-device_key-functionality | K82F, K64F | ARM, GCC_ARM, IAR
features-storage-tests-kvstore-static_tests | K82F, K64F | ARM, GCC_ARM, IAR
features-storage-tests-kvstore-general_tests_phase_1 | K82F, K64F | ARM, GCC_ARM, IAR
features-storage-tests-kvstore-securestore_whitebox | K82F, K64F | ARM, GCC_ARM, IAR
tests-psa-prot_internal_storage | K64F | ARM, GCC_ARM, IAR

I tried to run the tests on master branch (`K64F/ARM`) and compare the results with and without `SD` component which is now disabled for `K64F` by this PR. I got the results consistent with above failures:


Test | SD | FLASHIAP | Result
-- | -- | -- | --
features-storage-tests-blockdevice-general_block_device | yes | yes | failed*
features-device_key-tests-device_key-functionality | yes | yes | passed
features-storage-tests-kvstore-static_tests | yes | yes | passed
features-storage-tests-kvstore-general_tests_phase_1 | yes | yes | passed
features-storage-tests-kvstore-securestore_whitebox | yes | yes | passed
tests-psa-prot_internal_storage | yes | yes | passed
  |   |   |  
features-storage-tests-blockdevice-general_block_device | no | yes | failed
features-device_key-tests-device_key-functionality | no | yes | failed
features-storage-tests-kvstore-static_tests | no | yes | failed
features-storage-tests-kvstore-general_tests_phase_1 | no | yes | failed
features-storage-tests-kvstore-securestore_whitebox | no | yes | failed
tests-psa-prot_internal_storage | no | yes | failed
  |   |   |  
features-storage-tests-blockdevice-general_block_device | no | no | passed
features-device_key-tests-device_key-functionality | no | no | skipped
features-storage-tests-kvstore-static_tests | no | no | passed
features-storage-tests-kvstore-general_tests_phase_1 | no | no | passed
features-storage-tests-kvstore-securestore_whitebox | no | no | skipped
tests-psa-prot_internal_storage | no | no | passed**

`*` - `features-storage-tests-blockdevice-general_block_device` test has failed on master, but I checked that it regularly passes the CI test runs. Maybe there is some SD card issue on my side.

`**` - also PSA support needs to be disabled since it requires storage.

**Conclusion:**
It looks like there is a general problem on `K64F` (and `K82F`) with storage tests when SD Card support is disabled and Flash is used. I will create issue for this problem, but these failures are not related to spi feature branch, so I suggest do disable also `FLASHIAP` for now. Above results proves that this should solve the problem. @donatieng Fix can be found here: PR https://github.com/ithinuel/mbed-os/pull/8.

donatieng commented:
Thanks @mprse for the detailed investigation, we'll have to track the storage issue.

@ARMmbed/mbed-os-maintainers this is now ready for CI (again ;))

0xc0170 commented:
CI started

donatieng commented:
Woohoo!

ithinuel commented:
🕺
